perf: cache debug images

kevinrenskers opened the pull request:
## :scroll: Description

We're now caching the list of debug images, and invalidate the cache with the `_dyld_register_func_for_add_image` and `_dyld_register_func_for_remove_image` hooks.

#skip-changelog

## :bulb: Motivation and Context

Closes #1096

## :green_heart: How did you test it?

Unit test

## :pencil: Checklist

<!--- Put an `x` in the boxes that apply -->

- [x] I reviewed the submitted code
- [x] I added tests to verify the changes
- [x] I updated the docs if needed
- [x] Review from the native team if needed
- [x] No breaking changes

## :crystal_ball: Next steps

kevinrenskers commented:
I feel kinda lost with this ticket, as I know very very little about C code, static/dynamic libraries, or even how to see these in the Sentry UI. And how would we test this? Do we load a dynamic library in the unit tests? How?

I'm sorry but this is a bit above my knowledge level, and I am going to need some help.

brustolin commented:
For tests you create a mock SentryCrashImage provider with test data. and call the same methods `_dyld_register_func_for_add_image` and `_dyld_register_func_for_remove_image` would call to see whether `SentryDebugimageProvider` recorded the changes.

kevinrenskers commented:
> For tests you create a mock SentryCrashImage provider with test data. and call the same methods `_dyld_register_func_for_add_image` and `_dyld_register_func_for_remove_image` would call to see whether `SentryDebugimageProvider` recorded the changes.

Right, of course. But I mean how do we make sure those 2 hooks are actually getting called, and clearing the cache? Mocking it and calling it ourself doesn't test that.

philipphofmann commented:
> > For tests you create a mock SentryCrashImage provider with test data. and call the same methods `_dyld_register_func_for_add_image` and `_dyld_register_func_for_remove_image` would call to see whether `SentryDebugimageProvider` recorded the changes.
> 
> Right, of course. But I mean how do we make sure those 2 hooks are actually getting called, and clearing the cache? Mocking it and calling it ourself doesn't test that.

I think it is fine to test once manually if these hooks get actually called with manually loading a lib during the runtime, maybe [this](https://github.com/patriknyblad/ios-runtime-loading-dynamic-framework) works. Then we can keep the mocks for the rest. Ideally, we could load and unload libs during the tests, we could use [dlopen](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/dlopen.3.html#//apple_ref/doc/man/3/dlopen) and [dlclose](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man3/dlclose.3.html) for this. Maybe this [example helps](https://gist.github.com/tailriver/30bf0c943325330b7b6a).

indragiek commented on Sources/Sentry/SentryDebugImageProvider.m:
```diff
@@ -36,9 +39,24 @@ - (instancetype)initWithBinaryImageProvider:(id<SentryCrashBinaryImageProvider>)
         self.binaryImageProvider = binaryImageProvider;
     }
+
+    _dyld_register_func_for_add_image(clearCachedDebugImages);
+    _dyld_register_func_for_remove_image(clearCachedDebugImages);
+
```
In some cases your callback function may be called on a different thread (namely the thread that is loading the library), see here: https://twitter.com/gparker/status/1266572192126386176?s=20&t=jXOtWHAPuZ8oUrlUj788jQ

In this scenario the callback functions are accessing `cachedDebugImages` in a non-thread safe way, so you would need to use a lock to guard all reads/writes to `cachedDebugImages`.

brustolin commented:
What do we need to finish this?

kevinrenskers commented:
Philip said that basically he was going to take over this PR because it was going to be part of bigger changes. I'm not really sure what his idea was, and if we can still continue with this and get it merged (after addressing the feedback).

brustolin commented:
Oh, ok. 
I remember, we have a bigger problem regarding this images being fetch during a crash.

philipphofmann commented:
I think we should revisit this PR after fixing https://github.com/getsentry/sentry-cocoa/issues/1892, as both issues will touch similar code, and https://github.com/getsentry/sentry-cocoa/issues/1892 has a higher priority. So basically this PR is blocked by https://github.com/getsentry/sentry-cocoa/issues/1892.
